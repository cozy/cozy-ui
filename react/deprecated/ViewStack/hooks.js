import { useState } from 'react'

const sleep = duration => new Promise(resolve => setTimeout(resolve, duration))

const useStack = (initialState, options = {}) => {
  const [arr, setArray] = useState(initialState)
  const [curIndex, setCurIndex] = useState(initialState.length - 1)

  const push = (item, itemOptions) => {
    const newArr = [...arr, itemOptions ? [item, itemOptions] : item]
    setArray(newArr)
    setCurIndex(curIndex + 1)
  }

  const pop = async () => {
    if (arr.length <= 1) {
      return
    }
    const newArr = arr.slice(0, -1)
    setCurIndex(curIndex - 1)
    if (options.popDelay) {
      await sleep(options.popDelay)
    }
    setArray(newArr)
  }

  return [arr, curIndex, push, pop]
}

export { useStack }
