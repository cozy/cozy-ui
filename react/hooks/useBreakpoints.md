## useBreakpoints

Used to have access to the current breakpoint based on the window's size.
The component will be refreshed if the window's size changes.

```jsx static
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

const Component = () => {
  const { isMobile } = useBreakpoints()
}
```
