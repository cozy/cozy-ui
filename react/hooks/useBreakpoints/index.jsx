import React, { createContext, useContext, useState, useEffect } from 'react'
import throttle from 'lodash/throttle'
import breakpointDefs, {
  getBreakpointsStatus as _getBreakpointsStatus
} from '../../helpers/breakpoints'

const getBreakpointsStatus = () => _getBreakpointsStatus(breakpointDefs)

const BreakpointsCtx = createContext(null)

export const BreakpointsProvider = ({ children }) => {
  const [breakpoints, setBreakpoints] = useState(getBreakpointsStatus())
  useEffect(() => {
    const handleResize = throttle(() => {
      setBreakpoints(getBreakpointsStatus())
    }, 100)
    window.addEventListener('resize', handleResize)
    return () => {
      window.removeEventListener('resize', handleResize)
    }
  }, [])
  return (
    <BreakpointsCtx.Provider value={breakpoints}>
      {children}
    </BreakpointsCtx.Provider>
  )
}

const useBreakpoints = () => {
  const v = useContext(BreakpointsCtx)
  if (v === null) {
    throw new Error(
      'Cannot use useBreakpoints without BreakpointsProvider. The component must have a BreakpointsProvider above it.'
    )
  }
  return v
}

export default useBreakpoints
