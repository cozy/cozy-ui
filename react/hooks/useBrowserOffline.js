import { useState, useCallback } from 'react'
import useEventListener from './useEventListener'

/**
 * Checks if the browser is offline
 *
 * This is done through the browser API. It cannot garantees that the browser
 * is able to contact the serveur, nor that the connection is stable enough.
 *
 * In Nodejs, where this API doesn't exists, always return `false`.
 * @returns {boolean}
 */
export default function useBrowserOffline() {
  function isOfflineNow() {
    return window && window.navigator && window.navigator.onLine === false
  }

  const [isOffline, setOfflineSince] = useState(isOfflineNow)

  const setNavigatorOffline = useCallback(() => setOfflineSince(new Date()), [
    setOfflineSince
  ])
  const setNavigatorOnline = useCallback(() => setOfflineSince(false), [
    setOfflineSince
  ])

  useEventListener(window, 'online', setNavigatorOnline)
  useEventListener(window, 'offline', setNavigatorOffline)

  return !!isOffline
}
