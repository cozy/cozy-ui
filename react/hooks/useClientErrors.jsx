import filter from 'lodash/filter'
import React, { useCallback, useState, useEffect } from 'react'

import { useClient } from 'cozy-client'
import { FetchError } from 'cozy-stack-client'

import QuotaAlert from '../deprecated/QuotaAlert'

/**
 * Rendering functions for client fetch errors by status code
 *
 * @private
 * @type {object}
 */
const byHttpStatus = {
  413: QuotaError // <QuotaError />
}

/**
 * Display for a quota error from the client
 *
 * @see QuotaAlert
 * @private
 * @param {object} props - Props
 * @param {Function} props.dismiss - remove the error from the stack to display
 * @returns {React.ReactElement}
 */
function QuotaError({ dismiss }) {
  return <QuotaAlert onClose={dismiss} />
}

/**
 * Returns the handler for an error
 *
 * @param {import("../types").ClientError} error - The error
 * @returns {Function|null} React Component
 */
function getErrorComponent(error) {
  if (error instanceof Response || error instanceof FetchError) {
    const status = error.status || ''
    return byHttpStatus[status] || null
  }
  return null
}

/**
 * Renders a stack of errors
 *
 * @private
 * @see ClientErrors
 * @param {import("../types").ClientError[]} errorStack - array of errors/exceptions
 * @param {Function} setErrorStack - mutates the array of errors
 * @returns {Array<React.ReactElement>} React rendering
 */
function renderErrors(errorStack, setErrorStack) {
  const errors = errorStack.map((error, key) => {
    const Component = getErrorComponent(error)
    if (Component) {
      const dismiss = () =>
        setErrorStack(stack => filter(stack, e => e !== error))
      return <Component key={key} error={error} dismiss={dismiss} />
    } else {
      return null
    }
  })
  return errors
}

/**
 * Manages the client errors and allows to display them
 *
 * Returns a `ClientErrors` React component that takes care
 * of how to display cozy-client errors (probably displaying a modal)
 *
 * Only Quota Errors are managed for now.
 *
 * @example
 * ```
 * const App = () => {
 *   const { ClientErrors } = useClientErrors()
 *
 *   return <Layout>
 *      <h1>My app</h1>
 *      <ClientErrors />
 *   </Layout>
 * }
 * ```
 *
 * @param {object} [props] - Props
 * @param {boolean} [props.handleExceptions] - should cozy-client directly handle errors before forwarding them to the caller?
 * @returns {{ClientErrors: Function}} React component
 */
export default function useClientErrors({ handleExceptions = true } = {}) {
  const client = useClient()
  const [errorStack, setErrorStack] = useState([])

  /**
   * Handle client errors, add them to the error stack
   *
   * @param {import("../types").ClientError} error -
   * @returns {boolean} true if the error was manager, false otherwise
   */
  const handleError = useCallback(
    error => {
      // `isErrorManaged` is there to avoid the same error to be added twice
      // once the error has been added once, the `isErrorManaged`is set to true
      // and any future push is ignored
      if (error.isErrorManaged) return false

      const isManageable = !!getErrorComponent(error)
      if (isManageable) {
        error.isErrorManaged = true
        setErrorStack(stack => stack.concat(error))
        return true
      } else {
        error.isErrorManaged = false
        return false
      }
    },
    [setErrorStack]
  )

  useEffect(() => {
    if (handleExceptions) {
      client.on('error', handleError)
      return () => client.removeListener('error', handleError)
    } else {
      return undefined
    }
  }, [client, handleError, handleExceptions])

  // @ts-ignore
  const ClientErrors = useCallback(
    () => renderErrors(errorStack, setErrorStack),
    [errorStack, setErrorStack]
  )
  // @ts-ignore
  ClientErrors.displayName = 'ClientErrors'
  return { ClientErrors }
}
