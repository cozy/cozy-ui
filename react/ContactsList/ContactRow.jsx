import React from 'react'
import cx from 'classnames'
import PropTypes from 'prop-types'

import { Contact } from 'cozy-doctypes'
import { models } from 'cozy-client'

import ListItem from '../MuiCozyTheme/ListItem'
import ContactPhone from './Contacts/ContactPhone'
import ContactIdentity from './Contacts/ContactIdentity'
import ContactCozy from './Contacts/ContactCozy'
import ContactEmail from './Contacts/ContactEmail'
import useBreakpoints from '../hooks/useBreakpoints'

const { getPrimaryCozy, getPrimaryPhone, getPrimaryEmail } = models.contact

const ContactRow = ({ className, contact, onClick, divider, ...rest }) => {
  const { isMobile } = useBreakpoints()
  const phone = getPrimaryPhone(contact) || undefined
  const email = getPrimaryEmail(contact) || undefined
  const cozyUrl = getPrimaryCozy(contact) || undefined

  return (
    <ListItem
      data-testid="contact-listItem"
      className={cx({ 'u-c-pointer': onClick }, className)}
      divider={divider}
      onClick={() => onClick(contact)}
      {...rest}
    >
      <ContactIdentity contact={contact} />
      {!isMobile && <ContactEmail email={email} />}
      {!isMobile && <ContactPhone phone={phone} />}
      {!isMobile && <ContactCozy cozyUrl={cozyUrl} />}
    </ListItem>
  )
}

ContactRow.propTypes = {
  contact: Contact.propType.isRequired,
  onClick: PropTypes.func,
  divider: PropTypes.bool
}

export default ContactRow
