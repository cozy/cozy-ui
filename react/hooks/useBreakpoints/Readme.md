You can use the `useBreakpoints` hook to know whether
the window has a mobile or desktop width.

A `BreakpointsProvider` must be used above the component using the
`useBreakpoints` hook to provide it with the breakpoints' context.

When resizing the window, the components using `useBreakpoints` will
automatically update. Components below `BreakpointsProvider` but not
using `useBreakpoints` will not re-render (you can resize the window
with the inspector open to see that "render wrapper div" messages appear
only once).

The difference with media-queries is that you can use useBreakpoints to
completely change the content, the layout, or prune away complete
subtrees (useful on mobile where space is at a premium). Here for
example, the square contents changes on desktop or mobile.

```jsx
import useBreakpoints, { BreakpointsProvider } from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

const styles = {
  responsiveDiv: {
    color: 'white',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    margin: '0.5rem 0.5rem'
  }
}

const ResponsiveDiv = () => {
    const { isMobile } = useBreakpoints()
    return <div style={{
        ...styles.responsiveDiv,
        background: isMobile ? 'crimson' : 'royalblue',
        width: isMobile ? 100 : 200,
        height: isMobile ? 100 : 200
    }}>
      { isMobile ? 'mobile' : 'desktop' }
    </div>
};

const LoggingWrapper = ({ children, n }) => {
  console.log('render wrapper div', n)
  return children
}

<BreakpointsProvider>
  <LoggingWrapper n={1}>
    <LoggingWrapper n={2}>
      <ResponsiveDiv />
    </LoggingWrapper>
  </LoggingWrapper>
  <LoggingWrapper n={3}>
    <LoggingWrapper n={4}>
      <LoggingWrapper n={5}>
        <ResponsiveDiv />
      </LoggingWrapper>
    </LoggingWrapper>
  </LoggingWrapper>
</BreakpointsProvider>
```
