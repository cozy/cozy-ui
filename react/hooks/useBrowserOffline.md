Access the current network status.

```jsx static
import useBrowserOffline from 'cozy-ui/transpiled/react/hooks/useBrowserOffline'

const Component = () => {
  const isOffline = useBrowserOffline()
}
```
