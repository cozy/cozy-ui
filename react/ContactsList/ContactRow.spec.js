import React from 'react'
import { mount } from 'enzyme'

import { BreakpointsProvider } from '../hooks/useBreakpoints'
import ContactRow from './ContactRow'
import renderer from 'react-test-renderer'
import logger from 'cozy-logger'

logger.setLevel('error')

const makeContact = attrs => ({
  _id: 'contactid',
  _type: 'io.cozy.contacts',
  ...attrs
})

const setup = ({ contact }) => {
  const root = mount(
    <BreakpointsProvider>
      <ContactRow contact={contact} />
    </BreakpointsProvider>
  )
  return { root }
}

describe('ContactRow', () => {
  it('should accept the strict minimum', () => {
    const contact = makeContact({ email: [{ address: 'johndoe@localhost' }] })
    const { root } = setup({ contact })
    const contactrowemail = root.find('ContactEmail')
    expect(contactrowemail.text()).toBe(contact.email[0].address)
  })

  it('should display data', () => {
    const contact = makeContact({
      name: { familyName: 'Doe', givenName: 'John' },
      phone: [{ number: '0123456789' }],
      email: [{ address: 'johndoe@localhost' }],
      cozy: [{ url: 'http://johndoe.mycozy.cloud' }]
    })
    const { root } = setup({ contact })
    const contactrowname = root.find('ContactName')
    expect(contactrowname).toBeDefined()
    expect(contactrowname.text()).toEqual(
      expect.stringContaining(contact.name.givenName)
    )
    expect(contactrowname.text()).toEqual(
      expect.stringContaining(contact.name.familyName)
    )
    const contactrowphone = root.find('ContactPhone')
    expect(contactrowphone).toBeDefined()
    expect(contactrowphone.text()).toBe(contact.phone[0].number)

    const contactrowcozyurl = root.find('ContactCozy')
    expect(contactrowcozyurl).toBeDefined()
    expect(contactrowcozyurl.text()).toBe(contact.cozy[0].url)
  })

  it('should display email for missing information', () => {
    const contact = makeContact({
      email: [{ address: 'johndoe@localhost' }]
    })
    const { root } = setup({ contact })
    const contactrowname = root.find('ContactName')
    expect(contactrowname).toBeDefined()
    expect(contactrowname.text().trim()).toBe('johndoe@localhost')
    const contactrowphone = root.find('ContactPhone')
    expect(contactrowphone).toBeDefined()
    expect(contactrowphone.text().trim()).toBe('—')
    const contactrowcozyurl = root.find('ContactCozy')
    expect(contactrowcozyurl).toBeDefined()
    expect(contactrowcozyurl.text().trim()).toBe('—')
  })

  it('should accept empty array', () => {
    const contact = makeContact({ email: [] })
    const { root } = setup({ contact })
    const contactrowemail = root.find('ContactEmail')
    expect(contactrowemail).toBeDefined()
    expect(contactrowemail.text()).toBe('—')
  })

  it('should match the contact snapshot', () => {
    const contact = makeContact({
      name: { familyName: 'Doe', givenName: 'John' },
      phone: [{ number: '0123456789' }],
      email: [{ address: 'johndoe@localhost' }],
      cozy: [{ url: 'http://johndoe.mycozy.cloud' }]
    })
    const tree = renderer
      .create(
        <BreakpointsProvider>
          <ContactRow contact={contact} />
        </BreakpointsProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
