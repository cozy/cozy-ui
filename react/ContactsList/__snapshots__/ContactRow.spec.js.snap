// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactRow should match the contact snapshot 1`] = `
<div
  className="styles__contact___169nD"
  onClick={[Function]}
>
  <div
    className="styles__contact-identity___mL3IJ"
  >
    <div
      className="styles__c-avatar___PpDI- styles__c-avatar--small___1Y_Pv styles__contact-avatar___3lZPs"
      style={
        Object {
          "backgroundColor": "var(--brightSun)",
          "color": "white",
        }
      }
    >
      
      <span
        className="styles__c-avatar-initials___310qC"
      >
        JD
      </span>
    </div>
    <div>
      <span
        className="styles__contact-firstname___2GPEr"
      >
        John
      </span>
      Â 
      <span
        className="styles__contact-lastname___134MM"
      >
        Doe
      </span>
    </div>
  </div>
  <div
    className="styles__contact-email___3n3q2"
  >
    johndoe@localhost
  </div>
  <div
    className="styles__contact-phone___1sA_m"
  >
    0123456789
  </div>
  <div
    className="styles__contact-cozyurl___3kBp5"
  >
    http://johndoe.mycozy.cloud
  </div>
</div>
`;
